ui/components/app/snaps/snap-ui-renderer/snap-ui-renderer.js

```diff
@@ -53,11 +89,86 @@ const SnapUIRendererComponent = ({
     // We do this to avoid useless re-renders.
     (oldState, newState) => isEqual(oldState.content, newState.content),
   );
-  const rawContent = interfaceState?.content;
-  const content =
-    rawContent?.type === 'Container' || !rawContent
+
+  const content = useMemo(() => {
+    const rawContent = interfaceState?.content;
+    return rawContent?.type === 'Container' || !rawContent
       ? rawContent
       : Container({ children: rawContent });
+  }, [interfaceState?.content]);
+
+  const requireScroll = useMemo(
+    () => Boolean(content?.props?.children?.[1]?.props?.requireScroll),
+    [content?.props?.children],
+  );
+
+  const isProgrammaticScrollRef = useRef(false);
+
+  const [scrollState, dispatch] = useReducer(scrollReducer, {
+    isProgrammaticScroll: false,
+    buttonsEnabled: !requireScroll,
+    showArrow: false,
+    hasReachedBottom: !requireScroll,
+  });
+
+  const { onScroll, scrollToBottom, isScrollable, ref } = useScrollRequired(
+    [],
+    {
+      // Only enable the hook if we need to handle scroll events
+      enabled: requireScroll,
+      // This lets us batch state updates and avoid an unnecessary render
+      onMeasure: ({ isScrollable: canScroll, hasMeasured }) => {
+        if (!requireScroll || !hasMeasured) {
+          return;
+        }
+        dispatch({
+          type: 'MANUAL_SCROLL',
+          isScrollable: canScroll,
+          isAtBottom: !canScroll,
+        });
+      },
+    },
+  );
+
+  const buttonsEnabled = useMemo(
+    () => (requireScroll ? scrollState.buttonsEnabled : true),
+    [requireScroll, scrollState.buttonsEnabled],
+  );
+
+  const handleScroll = useCallback(
+    (e) => {
+      if (!requireScroll) {
+        return;
+      }
+
+      const isActuallyAtBottom =
```

Comment:
Is this different from `useScrollRequired`'s `hasScrolledToBottomState`? Can we deduplicate?

Reply:
Yes, I don't think that hook should even export `isScrolledToBottom` or `hasScrolledToBottomState` tbh, that calculation should happen in the component itself because it sets state too early in the hook.

Reply:
I'm not sure I understand. It seems like we're duplicating a lot of logic here, can we update the hook instead?

Reply:
I'm still trying to understand why we actually need all this code. The `useScrollRequired` hook has a property that indicates if the user has scrolled to the bottom already, which seems to work fine in existing implementations. Can you explain why all this code is necessary, and why we can't simply rely on `useScrollRequired`?

Reply:
Hmm so I was definitely experiencing flickering issues before, it was most likely something else that was causing it. I re-factored to just use the hook: [355fe1e](https://github.com/MetaMask/metamask-extension/pull/29863/commits/355fe1eda591d3a04f306e6492da2282140ecdc9). Updating the tests in another commit.